jwst/pipeline/calwebb_spec2.py

```diff
@@ -201,80 +206,23 @@ def process_exposure_product(
                 else:
                     raise RuntimeError('Cannot determine WCS.')
 
-        # Do background processing, if necessary
-        if exp_type in WFSS_TYPES or len(members_by_type['background']) > 0:
-
-            if exp_type in WFSS_TYPES:
-                bkg_list = []           # will be overwritten by the step
-            else:
-                bkg_list = members_by_type['background']
-
-            # Setup for saving
-            self.bkg_subtract.suffix = 'bsub'
-            if multi_int:
-                self.bkg_subtract.suffix = 'bsubints'
-
-            # Backwards compatibility
-            if self.save_bsub:
-                self.bkg_subtract.save_results = True
-
-            # Call the background subtraction step
-            input = self.bkg_subtract(input, bkg_list)
-
-        # Apply NIRSpec MSA imprint subtraction
-        # Technically there should be just one.
-        # We'll just get the first one found
-        imprint = members_by_type['imprint']
-        if exp_type in ['NRS_MSASPEC', 'NRS_IFU'] and \
-           len(imprint) > 0:
-            if len(imprint) > 1:
-                self.log.warning('Wrong number of imprint members')
-            imprint = imprint[0]
-            input = self.imprint_subtract(input, imprint)
-
-        # Apply NIRSpec MSA bad shutter flagging
-        if exp_type in ['NRS_MSASPEC', 'NRS_IFU']:
-            input = self.msa_flagging(input)
+        # Steps whose order is the same for all types of input.
+        input = self.bkg_subtract(input, members_by_type['background'])
+        input = self.imprint_subtract(input, members_by_type['imprint'])
+        input = self.msa_flagging(input)
 
         # The order of the next few steps is tricky, depending on mode:
         # WFSS/Grism data need flat_field before extract_2d, but other modes
         # need extract_2d first. Furthermore, NIRSpec MOS and FS need
         # srctype and wavecorr before flat_field.
         if exp_type in ['NRC_WFSS', 'NIS_WFSS', 'NRC_TSGRISM']:
+            input = self._process_grism(input)
             # Apply flat-field correction
-            input = self.flat_field(input)
-            input = self.extract_2d(input)
-            input = self.srctype(input)
+        elif exp_type in ['NRS_FIXEDSLIT', 'NRS_BRIGHTOBJ', 'NRS_MSASPEC', 'NRS_LAMP']:
+            input = self._process_nirspec(input)
```

Comment:
Technically this isn't a generic nirspec routine, because it doesn't apply to NRS_IFU. So maybe a somewhat less generic name for the method (_process_nirspec_slits?)

Reply:
Also, NRS_LAMP exposures are going to be covered by the pars reference files, and the set of steps to be run differs from that for regular NIRSPEC science data

Reply:
@stscirij Though the set of steps are different, is the order of those steps the same as what is done in `_process_nirspec`? The key between these methods is that ordering is different. If steps are skipped, that is OK and is expected.